homeassistant/components/pvpc_hourly_pricing/__init__.py

```diff
@@ -52,3 +73,54 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Unload a config entry."""
     return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Migrate old entry."""
+    if config_entry.version == 1:
+        _LOGGER.warning(
+            "Migrating PVPC sensor from old tariff '%s' to new '%s'. "
+            "Configure the integration to set your contracted power, "
+            "and select prices for Ceuta/Melilla, "
+            "if that is your case",
+            config_entry.data[ATTR_TARIFF],
+            _DEFAULT_TARIFF,
+        )
+        defaults = {
+            ATTR_TARIFF: _DEFAULT_TARIFF,
+            ATTR_POWER: DEFAULT_POWER_KW,
+            ATTR_POWER_P3: DEFAULT_POWER_KW,
+        }
+        config_entry.version = 2
```

Comment:
We try to avoid migrating the config entry version since that makes it impossible to role back for users. Can we detect the need to migrate by looking at the config entry data and just migrate the data as needed without changing config entry version?

Reply:
yes, of course, I'll change that 👍

Reply:
should I look at the `config_entry.options`, which will be empty for old entries, and fill it with the changes without increasing version?

Reply:
@MartinHjelmare, 

if  `VERSION = 1` remains in the ConfigFlow, then the `async_migrate_entry` is not called... did you suggest that I should avoid entirely the 'migrate-entry' flow and **just edit the entries inside the `async_setup_entry`** (in init or sensor)? I'm not sure if I understood correctly, sorry 😓

About this migration thing, I would also want to add a new unit test for that (coverage control is complaining about), but I'm having issues with the flow and the MockConfigEntry/MockEntity/MockEntityPlatform objects 😖. I know for sure that works in practice, but I think it should be reflected on tests too :)

Reply:
Yes, in `async_setup_entry` in the init module, ie here in this module, we can just use `hass.config_entries.async_update_entry` to update the entry as we need to.

Yes, migrations are very good to test. Let me know if I can give some input on something for that. The standard way is to add a MockConfigEntry with the old data, and test that it's updated to the correct data state after the integration is setup. Same for the entity registry migration, set up the integration from old data and let the migration happen and then assert the new entity registry state.

Reply:
I moved the migration logic inside the setup, and works as expected for _real_ sensors: first one gets migration and others get removed, but I cannot reproduce that logic with the MockConfigEntry, as the second call to `await async_migrate_entries(hass, entry.entry_id, update_unique_id)` does not fail with ValueError, and instead I get the "Platform XXX does not generate unique IDs" error :( 

I suppose it's because the test doesn't handle a proper entity registry with old sensors, but I'm not sure what I'm doing wrong 😅

The test I have is as follows:
```python
async def test_multi_sensor_migration(
    hass, caplog, legacy_patchable_time, pvpc_aioclient_mock: AiohttpClientMocker
):
    """Test tariff migration when there are >1 old sensors."""
    hass.config.time_zone = dt_util.get_time_zone("Europe/Madrid")
    uid_1 = "discrimination"
    uid_2 = "normal"
    old_conf_1 = {CONF_NAME: "test_pvpc_1", ATTR_TARIFF: uid_1}
    old_conf_2 = {CONF_NAME: "test_pvpc_2", ATTR_TARIFF: uid_2}

    config_entry_1 = MockConfigEntry(domain=DOMAIN, data=old_conf_1, unique_id=uid_1)
    config_entry_1.add_to_hass(hass)
    config_entry_2 = MockConfigEntry(domain=DOMAIN, data=old_conf_2, unique_id=uid_2)
    config_entry_2.add_to_hass(hass)

    mock_data = {"return_time": datetime(2019, 10, 27, 20, tzinfo=date_util.UTC)}

    def mock_now():
        return mock_data["return_time"]

    with patch("homeassistant.util.dt.utcnow", new=mock_now):
        await hass.config_entries.async_setup(config_entry_1.entry_id)

       # here it raises an OperationNotAllowed, but the 'unique IDs' error is already fired ¿?
       with pytest.raises(Exception):
            await hass.config_entries.async_setup(config_entry_2.entry_id)

        # await hass.config_entries.async_setup(config_entry_2.entry_id)
        assert_setup_component(1, DOMAIN)

        # check migration
        current_entries = hass.config_entries.async_entries(DOMAIN)
        # assert len(current_entries) == 2
        for migrated_entry in current_entries:
            assert migrated_entry.version == 1
            assert migrated_entry.data[ATTR_POWER] == migrated_entry.data[ATTR_POWER_P3]
            assert migrated_entry.data[ATTR_TARIFF] == TARIFFS[0]

        await hass.async_block_till_done()
```

Tomorrow I'll try to find an example to follow to define an entity registry to reflect a real migration + deletion for the scenario with >1 sensors

Reply:
> ~~Tomorrow I'll try to find an example to follow to define an entity registry to reflect a real migration + deletion for the scenario with >1 sensors~~

I think I got it 👍